### PR TITLE
3130 - 1

### PR DESCRIPTION
**Pull request by Shabanov-V**
## Номер
3130

## Номер задания
1

## Ссылка на видео с демо
https://youtu.be/Xlv2YFa1tQM

### CODE REVIEW

File: 3130/1/gem_info.rb [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+class GemInfo
+  WATCHER_WEIGHT = 0.5
+  USED_BY_WEIGHT = 1
+  STAR_WEIGHT = 0.7
+  CLOSE_ISSUE_WEIGHT = 0.9
+  OPEN_ISSUE_WEIGHT = -0.3
+
+  def initialize(gem_names)
+    @gem_names = gem_names
+    @gem_data = []
```
Comment (stahor) [inline 3130/1/gem_info.rb]: Not needed here. Make a method which returns this data. Use memoization if you need it more than once and the operation is slow

File: 3130/1/gem_info.rb [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  end
+
+  def call
+    fill_gem_data
+    sort_gems
+    @gem_data
```
Comment (stahor) [inline 3130/1/gem_info.rb]: call a method instead

File: 3130/1/gem_info.rb [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+      gem_repo = JSON.parse(
+        open("https://rubygems.org/api/v1/gems/#{gem_name}.json").read
+      )['source_code_uri'][%r{\/[^.:\/]+\/[^.:\/]+}]
+      github_info = JSON.parse(open("https://api.github.com/repos#{gem_repo}").read)
+      gem_stats = { name: gem_name }.merge(gem_statistic(gem_repo, github_info))
+      @gem_data << gem_stats
+    end
+  end
+
+  def gem_statistic(gem_repo, github_info)
```
Comment (stahor) [inline 3130/1/gem_info.rb]: make a new class for a single gem processing

File: 3130/1/gem_info.rb [hunk 4]
```diff
@@ -0,0 +1,84 @@
+require 'open-uri'
+require 'json'
+require 'nokogiri'
+
+class GemInfo
+  WATCHER_WEIGHT = 0.5
```
Comment (stahor) [inline 3130/1/gem_info.rb]: move them into YAML config file

File: 3130/1/top_gems.rb [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    generate_formatted_result
+  end
+
+  private
+
+  def parse_args(args)
+    hash_args = {}
+    args.each do |arg|
+      key, value = arg.scan(/--(\w+)=([^ ]+)/).flatten
+      hash_args.merge!(
```
Comment (stahor) [inline 3130/1/top_gems.rb]: weird formatting

File: 3130/1/top_gems.rb [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    @gem_data = GemInfo.new(@gem_names).call
+    count_column_sizes
+    generate_formatted_result
+  end
+
+  private
+
+  def parse_args(args)
+    hash_args = {}
+    args.each do |arg|
```
Comment (stahor) [inline 3130/1/top_gems.rb]: use inject

File: 3130/1/top_gems.rb (hunk 5 above)
Comment (stahor) [reply 3130/1/top_gems.rb]: why not just `hash_args[key.to_sym] = value` btw?

File: 3130/1/top_gems.rb [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+    @gem_data.first.keys.each do |key|
+      @col_size[key] = count_size_for_column(key)
+    end
+  end
+
+  def gem_statistic_value(gem_stats, key)
+    gem_stats[key].to_s.ljust(@col_size[key]).to_s
+  end
+
+  def format_gem(gem_stats)
```
Comment (stahor) [inline 3130/1/top_gems.rb]: there are gems for terminal tables actually. Maybe it's easier to use them instead. Not critical. I won't mind if u leave ur table

File: 3130/1/top_gems.rb [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+      " #{gem_statistic_value(gem_stats, :stargazers_count)} stars |" \
+      " #{gem_statistic_value(gem_stats, :forks_count)} forks |" \
+      " #{gem_statistic_value(gem_stats, :contributors)} contributors |" \
+      " #{gem_statistic_value(gem_stats, :open_issues)} issues |"
+  end
+
+  def generate_formatted_result
+    @gem_data = @gem_data.first(@hash_args[:top].to_i) if @hash_args[:top]
+    formatted_result = []
+    @gem_data.each do |gem_stats|
```
Comment (stahor) [inline 3130/1/top_gems.rb]: map

File: 3130/1/gem_info.rb [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+  def gem_score(gem)
+    gem[:used_by] * USED_BY_WEIGHT +
+      gem[:watchers_count] * WATCHER_WEIGHT +
+      gem[:stargazers_count] * STAR_WEIGHT +
+      gem[:close_issues] * CLOSE_ISSUE_WEIGHT +
+      gem[:open_issues] * OPEN_ISSUE_WEIGHT
+  end
+
+  def sort_gems
+    @gem_data = @gem_data.sort do |first, second|
```
Comment (stahor) [inline 3130/1/gem_info.rb]: use `sort_by` instead of `sort` + `reverse`

File: 3130/1/config.yml [hunk 10]
```diff
@@ -0,0 +1,9 @@
+gems:
+  - rails
+  - sinatra
+  - rspec
+  - devise
+  - jquery-rails
+  - rake
+  - reek
+  - rubocop
```
Comment (stahor) [inline 3130/1/config.yml]: newline

Comment (Shabanov-V) [reply 3130/1/config.yml]: @stahor Need review

**Pull request closed without merging**